### ISSUE TITLE
Warning `Task type ... declares property that is not annotated` even if property is annotated

### ISSUE THREAD

**Issue by koral--** [gradle/gradle#889]
### Expected Behavior
There is no warning about not annotated properties if they are annotated.

### Current Behavior
If annotated property (eg. having `@Internal`) is declared in base class then it is detected as not annotated in subclasses by task `validateTaskProperties` from `java-gradle-plugin`.
Eg. for example project linked below:
```
:buildSrc:validateTaskProperties
Task property validation finished with warnings:
  - Warning: Task type 'com.example.FooTask' declares property that is not annotated: 'foo$buildSrc_main'.
```

### Steps to Reproduce (for bugs)
Run any task (eg. `projects`) on this example project: https://github.com/koral--/gradle-java-plugin-internal-property-bogus-warning

### Your Environment
Build scan output does not contain messages from `buildSrc` project is it intended?

**Comment by koral--** [comment]
On Gradle 6.5 warning message is a little bit different but issue seems to be still valid:
```
   > Warning: Type 'BaseTask': field 'foo' without corresponding getter has been annotated with @Internal.
   > Warning: Type 'BaseTask': property 'foo$buildSrc' is not annotated with an input or output annotation.
   > Warning: Type 'FooTask': property 'foo$buildSrc' is not annotated with an input or output annotation.
```

**Comment by mlopatkin** [comment]
tl;dr - the workaround available now is to either use public properties or explicitly annotate only the getter with `@get:` syntax.

This is an interesting corner case. The `internal` visibility seems to be the culprit there. If the property is public or if it is annotated as `@get:Internal` then everything works fine. By default, Kotlin generates a backing private field named `foo` and a getter for it and applies annotation to the field. The exact JVM name of the getter depends on visibility though - the name for the getter of the `internal` field is mangled and becomes `getFoo$buildSrc`. Gradle is unable to determine the connection between field and getter and considers them separate "properties" causing all three warnings. We could try to do a better job there by examining Kotlin metadata - looks like it contains all the information needed to make the connection.

**Comment by ljacomet** [comment]
There is no plan to fix this automatically since putting this annotation on the getter works fine.
